### ISSUE TITLE
CNN PROJECT

### ISSUE THREAD

**Issue by Ridhwan36** [Ridhwan36/Machine-Learning-#2]
import pandas as pd
import re
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, Conv1D,
GlobalMaxPooling1D, Dense, Dropout
from tensorflow.keras.utils import to_categorical
# 1. Load dataset
df = pd.read_csv("/content/data.csv", on_bad_lines='skip', engine='python')
df = df[['title', 'source']].dropna()
df.columns = ['text', 'label']
# 2. Clean text
def clean(t):
t = re.sub(r'[^a-zA-Z\s]', '', str(t).lower())
return re.sub(r'\s+', ' ', t).strip()
df['text'] = df['text'].apply(clean)
# 3. Encode labels
le = LabelEncoder()
df['label'] = le.fit_transform(df['label'])
num_classes = len(le.classes_)
# 4. Split data
X_train, X_test, y_train, y_test = train_test_split(
df['text'], df['label'], test_size=0.2, stratify=df['label'], random_state=42)
# 5. Tokenize and pad
tok = Tokenizer(num_words=8000, oov_token='<OOV>')
tok.fit_on_texts(X_train)
X_train = pad_sequences(tok.texts_to_sequences(X_train), maxlen=150)
X_test = pad_sequences(tok.texts_to_sequences(X_test), maxlen=150)
y_train = to_categorical(y_train, num_classes)
y_test = to_categorical(y_test, num_classes)
# 6. Build CNN model
model = Sequential([
Embedding(8000, 100, input_length=150),
Conv1D(128, 5, activation='relu'),
GlobalMaxPooling1D(),
Dropout(0.4),
Dense(64, activation='relu'),
Dense(num_classes, activation='softmax')
])
model.compile(loss='categorical_crossentropy', optimizer='adam',
metrics=['accuracy'])
model.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.2,
verbose=1)
# 7. Evaluate and test
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print(f"\n✅ Test Accuracy: {acc*100:.2f}%")
sample = ["Jokowi meresmikan proyek baru di Jakarta"]
seq = pad_sequences(tok.texts_to_sequences(sample), maxlen=150)
pred = model.predict(seq)
print(f"🧠 Predicted Label: {le.inverse_transform([pred.argmax()])[0]}")
